### ISSUE TITLE
False positive unused_lifetimes warning

### ISSUE THREAD

**Issue by Hawk777** [rust-lang/rust#77217]
I tried this code:

```rust
#![warn(unused_lifetimes)]

pub struct Thingy<'x, X> {
	x: &'x mut X,
	y: &'static str,
}

impl<'x, X> Thingy<'x, X> {
	pub fn new(x: &'x mut X) -> Self {
		Self{x, y: "hello"}
	}

	pub async fn foo(&mut self) -> &'static str {
		self.y
	}
}
```

I got a warning “lifetime parameter `'x` never used”, on the `impl` line. But it clearly is used. The warning goes away if I delete the function `foo`, or make it a regular function instead of an asynchronous function.

### Meta
`rustc --version --verbose`:
```
rustc 1.46.0
binary: rustc
commit-hash: unknown
commit-date: unknown
host: x86_64-unknown-linux-gnu
release: 1.46.0
LLVM version: 10.0
```

I saw #61115 and rust-lang/rust-clippy#3988, but those were both closed at least a year ago and this is still happening in 1.46.0 which was just released recently, so I assume it’s not the same bug.

**Comment by lcnr** [comment]
The issue is that the field `x` is never marked as used, which is required to mark `'x` as used, as `x` is the only field using `'x`.

This is not ideal, but not sure on where exactly we want to lint here. If we have functions `f` and `g` with `f` calling `g` and `g` calling `f` it is probably good to lint for both `f` and `g` so I don't think just marking everything used in unused things as used is the right solution here.

**Comment by Hawk777** [comment]
Then why does the warning not show up if `foo` is not `async`? `x` would be equally unused in that case, but there's no warning.

**Comment by Hawk777** [comment]
Furthermore, this code, which clearly *does* use `x`, still has the unused lifetime warning, but of course not the unused field warning:
```rust
#![warn(unused_lifetimes)]

pub struct Thingy<'x> {
	x: &'x u32,
	y: &'static str,
}

impl<'x> Thingy<'x> {
	pub fn new(x: &'x u32) -> Self {
		Self{x, y: "hello"}
	}

	pub async fn foo(&self) -> &'static str {
		self.y
	}

	pub async fn bar(&self) -> u32 {
		*self.x
	}
}
```

Meanwhile, if I change both `async fn`s to `fn`, the warning goes away. It seems to be triggered by the presence of any `async fn` in the `impl`.

**Comment by lcnr** [comment]
Yeah, that is an actual bug :+1: thanks

**Comment by nikomatsakis** [comment]
So this is probably related to the desugaring somehow:

https://github.com/rust-lang/rust/blob/4d52dc4790cfaf3f23ff219c9a9932be2bfcd53e/compiler/rustc_ast_lowering/src/lib.rs#L1882-L1927

but I don't exactly know how yet. Maybe applying the desugaring to the example will help to see it.

**Comment by Hawk777** [comment]
I can also trigger a `single_use_lifetimes` warning with a free function:
```rust
#![warn(single_use_lifetimes)]
pub async fn foo<'a>(x: &'a u32) -> &'a u32 {
	x
}
```

Again, this only happens with an `async fn`, not with a regular `fn`, so I suspect it’s probably the same underlying cause (though if you disagree I’m happy to file a separate ticket).

**Comment by hyd-dev** [comment]
I also encountered this bug, and if I add `#[allow(unused_lifetimes)]` on the `async fn` (*instead of* the `impl`), the warning disappears, in both my code and [the code of this issue](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=633ed7166511e7c93a68214f5e55a94b).
I also read #78522. Is that the same bug? The warnings are both fired on the `async fn` itself.

**Comment by Vooblin** [comment]
@hyd-dev Yes, it is the same bug and I fixed it in this #79689 PR.

**Comment by tmandry** [comment]
Presumably fixed in #79689, leaving this open to try the original example in the next nightly.

**Comment by tmandry** [comment]
Triage: Confirmed that this is fixed in nightly.
